### DISCUSSION TITLE
Using puppet-editor-services with vimspector

### DISCUSSION THREAD

**Discussion by lelutin** [puppetlabs/community#55] [All Things Community]
### What Versions are you running?

OS Version: Debian unstable (sid)
VSCode Version: N/A (trying to integrate puppet-vscode with vimspector)
Puppet Extension Version: 1.0.0
PDK Version: N/A

I currently don't have PDK installed, so I'm using the puppet agent 5.5.22

I've downloaded the extension from github as the .vsix file, and extracted it.

### What You Are Seeing?

trying to start the extension with `node out/extension.js`, I get an error message about the `vscode` module not being found.

### What is Expected?

I'm trying to figure out if it's possible to use the puppet-vscode extension in Debian with the vimspector debugging IDE. If I could be able to run the extension code, then maybe vimspector would be able to interact with it.

https://github.com/puremourning/vimspector

### Output Log

To install this, I've first created a file for vimspector to injest. Then I get vimspector to download and extract the archive for me:

~~~
$ cat ~/.vim/bundle/vimspector/gadgets/custom/cust_puppet.json 
{
  "cust_vscode-puppet":  {
    "download": {
      "url": "https://github.com/puppetlabs/puppet-vscode/releases/download/${version}/${file_name}"
    },
    "all": {
      "version": "1.0.0",
      "file_name": "puppet-vscode-1.0.0.vsix",
      "checksum": "508e7971b4145909123e525d2da831218018cd39aeaa79f7f2438075d2ecd9dd",
      "adapters": {
        "cust_vscode-puppet": {
          "command": [
            "node",
            "${gadgetDir}/cust_vscode-puppet/out/extension.js"
          ],
          "name": "vscode-puppet",
          "configuration": {
            "cwd": "${workspaceRoot}",
            "showDebugOutput": false,
            "trace": false
          }
        }
      }
    }
  }
}
~/.vim/bundle/vimspector $ python3 install_gadget.py --update
Installing cust_vscode-puppet@1.0.0...
Checksum matches for /home/gabster/.vim/bundle/vimspector/gadgets/linux/download/cust_vscode-puppet/1.0.0/puppet-vscode-1.0.0.vsix, using it
Extracting /home/gabster/.vim/bundle/vimspector/gadgets/linux/download/cust_vscode-puppet/1.0.0/puppet-vscode-1.0.0.vsix to /home/gabster/.vim/bundle/vimspector/gadgets/linux/download/cust_vscode-puppet/1.0.0/root
Removing dir /home/gabster/.vim/bundle/vimspector/gadgets/linux/download/cust_vscode-puppet/1.0.0/root
Removed file /home/gabster/.vim/bundle/vimspector/gadgets/linux/cust_vscode-puppet
 - Done installing cust_vscode-puppet
Done
The following adapters were installed successfully:
 - cust_vscode-puppet
~~~

Then I try to execute the plugin to see if it's working:

~~~
~/.vim/bundle/vimspector/gadgets/linux/cust_vscode-puppet $ node out/extension.js 
internal/modules/cjs/loader.js:834
  throw err;
  ^

Error: Cannot find module 'vscode'
Require stack:
- /home/gabster/.vim/bundle/vimspector/gadgets/linux/download/cust_vscode-puppet/1.0.0/root/extension/out/extension.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:831:15)
    at Function.Module._load (internal/modules/cjs/loader.js:687:27)
    at Module.require (internal/modules/cjs/loader.js:903:19)
    at require (internal/modules/cjs/helpers.js:74:18)
    at Object.<anonymous> (/home/gabster/.vim/bundle/vimspector/gadgets/linux/download/cust_vscode-puppet/1.0.0/root/extension/out/extension.js:16:16)
    at Module._compile (internal/modules/cjs/loader.js:1015:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1035:10)
    at Module.load (internal/modules/cjs/loader.js:879:32)
    at Function.Module._load (internal/modules/cjs/loader.js:724:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:60:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/gabster/.vim/bundle/vimspector/gadgets/linux/download/cust_vscode-puppet/1.0.0/root/extension/out/extension.js'
  ]
}
~~~

Running npm install against the extracted .vsix gives this -- but this is probably expected, since the .vsix archive is supposed to be a pre-built version of the extension.

~~~
~/.vim/bundle/vimspector/gadgets/linux/cust_vscode-puppet $ npm install
npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! Found: eslint@7.17.0
npm ERR! node_modules/eslint
npm ERR!   dev eslint@"^7.0.0" from the root project
npm ERR!   peer eslint@">=3.14.1" from eslint-config-prettier@6.15.0
npm ERR!   node_modules/eslint-config-prettier
npm ERR!     dev eslint-config-prettier@"^6.11.0" from the root project
npm ERR!   2 more (eslint-plugin-prettier, @typescript-eslint/experimental-utils)
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer eslint@"^5.0.0 || ^6.0.0" from @typescript-eslint/eslint-plugin@2.34.0
npm ERR! node_modules/@typescript-eslint/eslint-plugin
npm ERR!   dev @typescript-eslint/eslint-plugin@"^2.33.0" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.
npm ERR! 
npm ERR! See /home/gabster/.npm/eresolve-report.txt for a full report.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/gabster/.npm/_logs/2021-01-03T03_49_18_748Z-debug.log
~~~

I've also tried cloning the extension from git and following the instructions in README_BUILD.md. `npm install` completes successfully, but then the document says to run `node node_modules/vscode/bin/install` but this file is not present in the repository.

**Comment by lelutin** [comment]
trying to fix the dependency issue by installing an older version of eslint I was able to get the install to go through:

~~~
$ npm install vscode eslint@6
~~~

and now I can see a `node_modules/vscode` directory. however, I'm still getting the same error message about the 'vscode' module not being found when I start extension.js

**Comment by lelutin** [comment]
running `node node_modules/vscode/bin/install` after the above comment gave me the error:

~~~
~~~

I found this that seems relevant: https://github.com/Microsoft/vscode/issues/2810#issuecomment-182209917

I've added this in the "scripts" section of the package.json file:

~~~
"postinstall": "node ./node_modules/vscode/bin/install"
~~~

and then when I run `npm install` again in the vscode-puppet directory I can see that it's running and finding a file to download:

~~~
$ npm install

> puppet-vscode@1.0.0 postinstall
> node ./node_modules/vscode/bin/install

Detected VS Code engine version: ^1.38.0
Found minimal version that qualifies engine range: 1.38.0
Fetching vscode.d.ts from: https://raw.githubusercontent.com/Microsoft/vscode/3db7e09f3b61f915d03bbfa58e258d6eee843f35/src/vs/vscode.d.ts
vscode.d.ts successfully installed!


removed 16 packages, and audited 326 packages in 3s

44 packages are looking for funding
  run `npm fund` for details

4 low severity vulnerabilities

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.
~~~

but the missing module 'vscode' error persists

**Comment by jpogran** [comment]
Thanks for reporting this @lelutin, it's great to see people trying to use our extension in new and different ways than we intended.

Looking at your output and the errors you are seeing come from trying to run a VS Code extension without VS Code. The `vscode` error you are seeing is Node telling you it can't find the `vscode` npm package. This package isn't part of the package.json because it's inside VS Code at runtime and not available externally, nor do I think it is it meant to be used externally. The closest you can get to running an extension without VS Code itself is with the vscode-test harness, but that is meant for unit/acceptance testing and not running the extension.

That being said, I reviewed the vimspector page and see most of the languages are command lines for the debuggers for the language being used. The Puppet language doesn't have a debugger, command line or otherwise to use like this, but the part of the Puppet Extension that provides a debugging experience is in the [Puppet Editor Services](https://github.com/puppetlabs/puppet-editor-services) project. You can see how VS Code calls it to debug a Puppet manifest [here](https://github.com/puppetlabs/puppet-vscode/blob/main/src/feature/DebuggingFeature.ts#L42-L57). In short, it starts a ruby process which responds to [LSP](https://langserver.org) requests from the VS Code Debugger. Since vimspector isn't a LSP client, I'm not sure this will help you very much, but that is how you would start the Puppet debugger we have in VS Code. An example of this working for vim/neovim for editing (not debugging) is at https://voxpupuli.org/blog/2019/04/08/puppet-lsp-vim.

**Comment by lelutin** [comment]
thanks! this reponse was super informative :)

I'll start with using that blog post to figure out how to use the language-server part, either with youcompleteme (which I currently use) as the LSP client or maybe switch over to coc.vim -- that should already be super useful in itself. I did not know about puppet-editor-services until now!

for the debugger part, if I undestand correctly: the protocol specific to debugging is named differently and it would be DAP (https://microsoft.github.io/debug-adapter-protocol/overview)? If that's correct, than maybe I might be in luck since I if I'm not mistaken vimspector is specifically a DAP client.

so from what I understand from this, I should switch my focus to puppet-editor-services since it seems to match more directly with what I need to use.

cheers!

**Comment by lelutin** [comment]
Wow! just to let you know: I ended up getting `puppet-debugserver` to interact properly with vimspector and was able to step through code as the catalog was being compiled. thanks again for the pointers, it really gave me the nudge in the right direction.

I'll need to cleanup all of my notes and cofiguration, but I intend to send some configuration snippets to the vimspector project soon so that others can start debugging puppet manifests with vim, too.

cheers!

**Comment by glennsarti** [comment]
🎉🎉🎉 Excellent

**Comment by jpogran** [comment]
@lelutin thats awesome! Would you be interested in contributing those notes as a doc on https://puppet-vscode.github.io or a blog post?

**Comment by lelutin** [comment]
@jpogran that would indeed make this visible to more vim users out there

fwiw I've already added this to the vimspector project: https://github.com/puremourning/vimspector/wiki/languages#puppet

I wonder how it would fit on puppet-vscode.gh.io specifically though since that tool, as you pointed out to me, is specifically the VS Code glue plugin. but maybe ppl tend to look at puppet-vscode more directly than puppet-editor-services? idk.

either way, I'll surely use that opportunity to write something up on my blog (that thing has been stagnating for too long). and if you think it would make sense to add something to puppet-vscode.gh.io, I don't mind republishing there, maybe in a format more suitable to the docs.

I'll update when I have something published on my blog. maybe tonight

**Comment by jpogran** [comment]
I'm going to move this issue to a discussion, and we can work out somewhere to publish.

We have an umbrella page for all of the components that make up these things, I was thinking it could go there as we have a couple of other editors to document, but haven't finished making that yet.

**Comment by lelutin** [comment]
wow sorry, I fell into a wormhole. coc.nvim is a massive thing and I jumped in there head first (thanks again for pointing out dhollinger's post, I made some huge changes to my vim setup with this. coc.nvim is pretty awesome :) )

depending on the format of the page, the section about vimspector + puppet-editor-services could be real quick. or I can write something that reads a bit more like a howto for setting things up with more detailed steps -- the latter will require more time to write of couse.

**Comment by jpogran** [comment]
@lelutin Apologies for the delay. A short document about how to setup vimspector and puppet-editor services would be a great start. If you could open a PR in the lingua-pupuli/docs repo and add a new markdown file [here](https://github.com/lingua-pupuli/docs/tree/main/editor-services/content), we can get started on something. The path should look like `docs/editor-services/content/en/docs/vimspector.md`. Don't worry about getting it perfect before submitting, it can be a work in progress.

For the longer form how-to, I can get you linked up with the people who handle our blog, so we can get something on the official blog. Please send me an email at `james.pogran at puppet.com` and we'll get started.



If you could put a pull request in the lingua-pupuli/docs repo that makes a new markdown file [here](https://github.com/lingua-pupuli/docs/tree/main/vscode/content/en/docs/Features) we can work on the content and wording. When things are complete, I can move it to the correct place in the umbrella project.
